Add: `permExactMatchStringTrigger(...)`, `permColorTrigger(...)` and `permLuaFunctionTrigger(...)` functions

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Adds code/LUA API functions to allow the creation of (permanent) "exact match trigger", "color trigger" *and "Lua function trigger"* types via Lua scripts/packages.

#### Motivation for adding to Mudlet
Consistency in API with other existing `permXxxxTrigger(...)` functions.

#### Other info (issues closed, discussion etc)
Theoretically there ought to also be function~s~ for ~"lua function" and~ "line spacer" type item based trigger~s~ - however ~the latter of these~ *this* is (probably) not helpful on its own which suggests that additional functions/functionality to add/modify/remove individual "items" to existing triggers is desirable. ~Conversely, a separate PR to add a `permLuaFunctionTrigger(...)` function is now on my TODO list~ *also included herein*!

The functions added by this PR have been modified before publication to accept the recently added (by myself) "multi-line delta" argument of PR #7493 - though these functions post-date that PR this fifth argument is also considered to be optional so the new functions behave as the existing ones do.

The new `permColorTrigger(...)` function takes a table of sub-tables as the third argument. Whilst the "color trigger" type currently only handles either/both a foreground and a background colour (stored as a string representation with those elements as a three digit number inside that string or a couple of "special" string values otherwise) I have medium term plans to extend the type to handle other text formatting details (even those that we currently detect but do not reproduce: bold, italic, underline, blink, etc.) I expect to store this as a series of words, one for each aspect, appended to that string and possibly preceded with the prefix "'NOT_'" or the symbol "'!'". So this can be added later to the Lua API function it will also accept - but issue a `[ WARN ] -` additional string arguments in the "sub-table" for each trigger "item" after zero, one (foreground only) or two (foreground and background). This is so that if a script/package targetting a later Mudlet version that does handle this is used with a version of Mudlet with the `permColorTrigger(...)` as of this PR then the user will get a clear warning that it likely won't work as expected. (Obviously the current and previous Mudlet versions will just barf on a usage of the non-existing - in them - `permColorTrigger(...)`!)

closes #1573

SlySven committed:
Add: permExactMatchStringTrigger(...) and permColorTrigger(...) functions
#### Brief overview of PR changes/additions
Adds code/LUA API functions to allow the creation of (permanent) "exact
match" and "color trigger" types via Lua scripts/packages.

#### Motivation for adding to Mudlet
Consistency in API with other existing permXxxxTrigger(...) functions.

#### Other info (issues closed, discussion etc)
Theoretically there ought to also be functions for "lua function", "line
spacer" type item based triggers - however the latter of those is
(probably) not helpful on their own which suggests that additional
functions/functionality to add/modify/remove individual "items" to existing
triggers is desirable. Conversely, a separate PR to add a
`permLuaFunctionTrigger(...)` function is now on my TODO list!

The functions added by this PR have been modified before publication to
accept the recently added (by myself) "multi-line delta" argument of PR
#7493 - though these functions post-date that PR this fifth argument is
also considered to be optional so the new functions behave as the existing
ones do.

The new `permColorTrigger(...)` function takes a table of sub-tables as
the third argument. Whilst the "color trigger" type currently only handles
either/both a foreground and a background colour (stored as a string
representation with those elements as a three digit number inside that
string or a couple of "special" string values otherwise) I have medium term
plans to extend the type to handle other text formatting details (even
those that we currently detect but do not reproduce: bold, italic,
underline, blink, etc.) I expect to store this as a series of words, one
for each aspect, appended to that string and possibly preceded with the
prefix "'NOT_'" or the symbol "'!'". So this can be added later to the
Lua API function it will also accept - but issue a `[ WARN ] -` additional
string arguments in the "sub-table" for each trigger "item" after zero, one
(foreground only) or two (foreground and background). This is so that if
a script/package targetting a later Mudlet version that does handle this
is used with a version of Mudlet with the `permColorTrigger(...)` as of
this PR then the user will get a clear warning that it likely won't work as
expected. (Obviously the current and previous Mudlet versions will just
barf on a usage of the non-existing - in them - `permColorTrigger(...)`!)

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add `permLuaFunctionTrigger(...)` as well

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

ZookaOnGit commented:
/refresh links

vadi2 committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

ZookaOnGit commented:
When creating a `permColorTrigger` with this command `lua permColorTrigger("colour test", "", {fg255}, "send('test')")` it creates a trigger folder in the root called `colour test`, not an actual trigger.

I'm also finding that I can't delete the triggers straight away.  I have to close the script editor, reopen it, and only then can I delete?

vadi2 committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven commented:
> When creating a `permColorTrigger` with this command `lua permColorTrigger("colour test", "", {fg255}, "send('test')")` it creates a trigger folder in the root called `colour test`, not an actual trigger.

`lua permColorTrigger("colour test", "", {fg255}, "send('test')")` looks suspect anywhere - I thought I made it so that the third argument needs to be a table of sub-tables - the inner one being one of the following forms:
* {123} - a number representing the ANSI foreground colour number
* {123, 234} - a number representing the ANSI foreground colour, followed by an ANSI background colour number
* {123, "text", ...} - a number representing the ANSI foreground colour number followed by one or more strings representing the presence (or absence) of a formatting detail like "italic" or "!underline" - which is pending me extending the colour trigger type to support those other details and which will currently put up a "[ XXXX ] - message about colour trigger of type not supported (yet) by this Mudlet version"
* {123, 234, "text", ...} - a number representing the ANSI foreground colour number followed by an ANSI background colour number followed by one or more strings representing the presence (or absence) of a formatting detail like "italic" or "!underline" - again not supported yet.
* {"text", ...} - one or more strings representing the presence (or absence) of a formatting detail like "italic" or "!underline" - again not supported yet.

As such I'd have thought you'd have triggered that warning because fg255 is not a number - unless it is a variable that represented one. As well as the numbers in the range 0 to 255 for the colour codes there are a couple of extra ones `-1` for the "ignore" colour - so that `{-1, 15}` can be used to specify a colour trigger to match on a bright white background whilst ignoring the foreground or `-2` for the "default" colour - so that `{-2,-2}` would trigger on any text that has NOT been coloured. Note that `{-1,-1}` is non-functional and will be silently dropped as an "item" (row in the resulting color trigger in the editor)

Does: 
`lua permColorTrigger("colour test", "", {{255}}, "send('test')")` work for you?

> I'm also finding that I can't delete the triggers straight away.  I have to close the script editor, reopen it, and only then can I delete?

Yeah, I've noticed some issues with the buttons (more than the "Delete" one, I can't recall) on the trigger editor not working the first time it is shown - if you toggle to a different Mudlet item type and switch back to the "Triggers" view things seem to work again. IIRC there was some code put in to prevent maybe the "Delete" button working for something (maybe the top level item, or "Modules"/"Package" {Brown} folders) but it sounds like something is not being initialised properly (guess we need a separate issue about this **unrelated** thing)...

ZookaOnGit commented:
> (guess we need a separate issue about this **unrelated** thing)...

#7413

ZookaOnGit commented:
> As such I'd have thought you'd have triggered that warning because fg255 is not a number - unless it is a variable that represented one.

No, it wasn't a variable.  I didn't have documentation so was just playing around with whatever stuck and what was allowed.

> Does: `lua permColorTrigger("colour test", "", {{255}}, "send('test')")` work for you?

Yes.

Your warning message has a typo, "implimented".

Commit:
Fix: correct spelling error

Signed-off-by: Stephen Lyons <stephen@clare>

Commit:
Revise: detect and handle some corner cases

This will ensure we do not create any colour trigger items with both the
foreground and background colours both set to the ignored value.

It will also handle the example in https://github.com/Mudlet/Mudlet/pull/7534#issuecomment-2533992006
so as to not create a trigger in that bogus (the argument three contains
a table which itself just contains a `nil` value)...

Signed-off-by: Stephen Lyons <stephen@clare>

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven committed:
Revert: remove changes made to match #7493 since that's been reverted #7624

This is so that the current `development` branch can be merged in.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

Conflicts resolved in:
* src/TLuaInterpreter.cpp
* src/TLuaInterpreterMudletObjects.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

SlySven committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

ZookaOnGit committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

ZookaOnGit committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

ZookaOnGit committed:
Merge branch 'development' into Add_permExactMatchStringTrigger_and_permColorTrigger_functions

ZookaOnGit commented:
this tests good, just need documentation

vadi2 commented:
Any chance of movement on this? The code is ready.

vadi2 committed:
Merge development into Add_permExactMatchStringTrigger_and_permColorTrigger_functions
